File: source/codes/leetcode/src/main/java/leetcode/L_84.go
package sample

func largestRectangleArea(heights []int) int {
	left := make([]int, len(heights))
	right := make([]int, len(heights))
	stack := []int{}
	for i := 0; i < len(heights); i++ {
		for len(stack) > 0 && heights[stack[len(stack)-1]] >= heights[i] {
			stack = stack[:len(stack)-1]
		}
		if len(stack) == 0 {
			left[i] = -1
		} else {
			left[i] = stack[len(stack)-1]
		}
		stack = append(stack, i)
	}

	stack = []int{}
	for i := len(heights) - 1; i >= 0; i-- {
		for len(stack) > 0 && heights[stack[len(stack)-1]] >= heights[i] {
			stack = stack[:len(stack)-1]
		}
		if len(stack) == 0 {
			right[i] = len(heights)
		} else {
			right[i] = stack[len(stack)-1]
		}
		stack = append(stack, i)
	}
	ans := heights[0]
	for i := 0; i < len(heights); i++ {
		ans = max(ans, (right[i]-left[i]-1)*heights[i])
	}
	return ans
}

File: assets/codes/leetcode/src/main/java/top100liked/L_84.java
package top100liked;

import java.util.Stack;

// TODO 多练
public class L_84 {
    public int largestRectangleArea(int[] heights) {
        int max = 0;
        Stack<Integer> stack = new Stack<>();

        int i = 0;
        for (; i < heights.length; i++) {
            if (stack.isEmpty() || heights[i] > heights[stack.peek()]) {
                stack.push(i);
                continue;
            }
            while (!stack.isEmpty() && heights[stack.peek()] > heights[i]) {
                int topIndex = stack.pop();
                int new_top_index = stack.isEmpty() ? -1 : stack.peek();
                int height = heights[topIndex];
                int width = i - new_top_index - 1;
                max = Math.max(max, width * height);
            }
            stack.push(i);
        }

        while (!stack.isEmpty()) {
            int topIndex = stack.pop();
            int new_top_index = stack.isEmpty() ? -1 : stack.peek();
            int height = heights[topIndex];
            int width = i - new_top_index - 1;
            max = Math.max(max, width * height);
        }

        return max;
    }

//    public int largestRectangleArea(int[] heights) {
//        int max = 0;
//        for (int i = 0; i < heights.length; i++) {
//            int j = i;
//            for (; j < heights.length && heights[j] >= heights[i]; j++) {
//
//            }
//            int k = i;
//            for (; k >= 0 && heights[k] >= heights[i]; k--) {
//
//            }
//            max = Math.max(max, (j - k - 1) * heights[i]);
//        }
//        return max;
//    }

    public static void main(String[] args) {
        L_84 l84 = new L_84();
//        System.out.println(l84.largestRectangleArea(new int[]{2, 1, 5, 6, 2, 3}));
        System.out.println(l84.largestRectangleArea(new int[]{2, 4}));
    }
}
